pkg/controllers/user/nodesyncer/nodetaints.go

```diff
@@ -0,0 +1,54 @@
+package nodesyncer
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	nodehelper "github.com/rancher/rancher/pkg/node"
+	"github.com/rancher/rancher/pkg/taints"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func (n *nodesSyncer) syncTaints(key string, obj *v3.Node) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+
+	if !v3.NodeConditionRegistered.IsTrue(obj) {
+		return obj, nil
+	}
+
+	if obj.Spec.UpdateTaintsFromAPI == nil {
+		return obj, nil
+	}
+	node, err := nodehelper.GetNodeForMachine(obj, n.nodeLister)
+	if err != nil {
+		return obj, err
+	}
+	toAdd, toDel := taints.GetToDiffTaints(node.Spec.Taints, obj.Spec.DesiredNodeTaints)
```

Comment:
It looks like the diff result is based on `DesiredNodeTaints `, which is cleaned up in L51. Then existing taints will become delete candidates in next sync when `DesiredNodeTaints ` is empty. Should we skip sync in empty desired taints

Reply:
It won't, because we update `UpdateTaintsFromAPI` to nil and we will skip the sync if `UpdateTaintsFromAPI` == nil.